examples/cosmos/lightclient/Lightclient.qnt

```diff
@@ -0,0 +1,569 @@
+    */ 
+    pure def ULTIMATE_HEIGHT = TARGET_HEIGHT + 1 
+ 
+    import Blockchain(
+        AllNodes = AllNodes, 
+        ULTIMATE_HEIGHT = ULTIMATE_HEIGHT, 
+        TRUSTING_PERIOD = TRUSTING_PERIOD
+        ) as BC from "./Blockchain"
+
+    /************************** Lite client ************************************/
+
+    /* the heights on which the light client is working */  
+    pure def HEIGHTS = TRUSTED_HEIGHT.to(TARGET_HEIGHT)
+
+    /* the control states of the lite client */ 
+    pure def States = Set( "working", "finishedSuccess", "finishedFailure" )
+
+    // The verification functions are implemented in the API
+    import LCVerificationApi(
+        TRUSTING_PERIOD = TRUSTING_PERIOD, 
+        CLOCK_DRIFT = CLOCK_DRIFT,
+        REAL_CLOCK_DRIFT = REAL_CLOCK_DRIFT,
+        FAULTY_RATIO = FAULTY_RATIO
+        ) as API from "./LCVerificationApi"
+
+
+    /*
+    Initial states of the light client.
+    Initially, only the trusted light block is present.
+    */
+    action LCInit: bool = all {
+        {
+            nondet tm = oneOf(Int)
+            all {
+                tm >= 0,
+                API::IsLocalClockWithinDrift(tm, BC::refClock),
+                API::SetLocalClock(tm)
+            }
+        },
+        state' = "working",
+        nextHeight' = TARGET_HEIGHT,
+        nprobes' = 0, // no tests have been done so far
+        {
+            val trustedBlock = BC::blockchain.get(TRUSTED_HEIGHT)
+            val trustedLightBlock = {header: trustedBlock, Commits: AllNodes}
+            all {
+                // initially, fetchedLightBlocks is a function of one element, i.e., TRUSTED_HEIGHT
+                fetchedLightBlocks' = Map(TRUSTED_HEIGHT -> trustedLightBlock),
+                // initially, lightBlockStatus is a function of one element, i.e., TRUSTED_HEIGHT
+                lightBlockStatus' = Map(TRUSTED_HEIGHT -> "StateVerified"),
+                // the latest verified block the the trusted block
+                latestVerified' = trustedLightBlock,
+                MonitorTransition(trustedLightBlock, trustedLightBlock, API::localClock, "SUCCESS")
+            }
+        }
+    }
+
+    /*
+    Block should contain a copy of the block from the reference chain,
+    with a matching commit.
+    */
+    def CopyLightBlockFromChain(block: LightBlock, height: int): bool = {
+        val ref = BC::blockchain.get(height)
+        val lastCommit = 
+            if(height < ULTIMATE_HEIGHT) BC::blockchain.get(height + 1).lastCommit 
+            // for the ultimate block, which we never use, as ULTIMATE_HEIGHT = TARGET_HEIGHT + 1
+            else  BC::blockchain.get(height).VS 
+        block == {header: ref, Commits: lastCommit}
+    }   
+
+    // Either the primary is correct and the block comes from the reference chain,
+    // or the block is produced by a faulty primary.
+    //
+    // [LCV-FUNC-FETCH.1::TLA.1]
+    def FetchLightBlockInto(block: LightBlock, height: int): bool =
```

Comment:
No idea what this spec does, I just transcribed the names as they were.

Reply:
This operator is essentially a predicate that checks, whether the primary (correct or faulty) could fetch `block` into its set of received blocks. It does not actually fetch anything, as the block is chosen non-deterministically, but it constrains the block. The terminology is indeed confusing, since this spec borrowed concepts from the actual light client implementation.